src/Core/OOShipRegistry.m

```diff
@@ -209,6 +209,12 @@ - (NSDictionary *) shipInfoForKey:(NSString *)key
 	return [_shipData objectForKey:key];
 }
 
+- (void) setShipInfoForKey:(NSString *)key with:(NSDictionary *)newShipData
+{
+	NSMutableDictionary *mutableDict = [_shipData mutableCopy];
+	[mutableDict setObject:OODeepCopy(newShipData) forKey:key];
+	_shipData = [mutableDict mutableCopy];
+}
```

Comment:
I think you need to release mutableDict after being done using it. Can't check myself right now so I will have to ask, can you try `NSMutableDictionary *mutableDict = [NSMutableDictionary dictionaryWithDictionary:_shipdata];`? I think in this case you will not need to release manually.

The above may not work, but if you can give it a try and it does work, I think it may be a slightly better way of handling it. The general rule is that anything created with a copy selector requires a manual release.

Reply:
Also, _shipData is a dictionary, Why is it assigned a mutableDictionary type at the last line of the method?

Reply:
Your suggestion worked.

I pretty much copy-pasted it from https://stackoverflow.com/a/29391936/6860744

Reply:
In order to avoid further memory leakage, I would recommend changing
```_shipData = [mutableDict mutableCopy];``` (points _shipData to a new memory location without releasing the previous one)
to
```DESTROY(_shipData);``` (releases memory held by _shipData)
```_shipData = [[NSDictionary dictionaryWithDictionary:mutableDict] retain];``` (Reassigns _shipData to new location and retains it so it won't be lost when the method exits).

I have built the PR to test it on my system and, although the change in _shipData clearly takes place, I am unable to see the results of the change in the game. This is probably because there is no instruction to reload the registry with the new changes anywhere. But it seems to me, going through OOShipRegistry.m, that any attempt to reload the ship information would cause a reset of the current _shipData before reading the information back from the game cache, so I am not really sure how this is supposed to practically work. @tsoj: Do you happen to have an example of an OXP actually using the changes made to _shipData as a result of this PR?

Reply:
Nevermind the above, I got it to work (silly me, should have realized better at which points in the game the registry values are meant to be read). The changes in the registry do work, so I think that with the proposed changes applied, we are more or less there.